srml/support/procedural/src/storage/impls.rs

```diff
@@ -0,0 +1,414 @@
+		let option_simple_1 = option_unwrap(is_option);
+		// make sure to use different prefix for head and elements.
+		let head_key = format!("head of {}", prefix);
+		let prefix = format!("{}", prefix);
+		let key_for = quote! {
+			(<Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::key_for(key))
+		};
+		let linkage = syn::Ident::new(&format!("Linkage{}", name), name.span());
+		let enumerator = syn::Ident::new(&format!("Enumerator{}", name), name.span());
+		let put_or_insert = quote! {
+			match linkage {
+				Some(linkage) => storage.put(key_for, &(val, linkage)),
+				None => <Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::insert(key, &val, storage),
+			}
+		};
+		let mutate_impl = if !type_infos.is_option {
+			put_or_insert
+		} else {
+			quote! {
+				match val {
+					Some(ref val) => #put_or_insert,
+					None => <Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::remove(key, storage),
+				}
+			}
+		};
+
+		// generator for linked map
+		quote! {
+			#[derive(Default, parity_codec_derive::Encode, parity_codec_derive::Decode)]
+			struct #linkage {
+				/// Previous element key in storage (None for the first element)
+				previous: Option<#kty>,
+				/// Next element key in storage (None for the last element)
+				next: Option<#kty>,
+			}
+
+			impl #linkage {
+				/// Update linkage when this element is removed.
+				///
+				/// Takes care of updating previous and next elements points
+				/// as well as updates head if the element is first or last.
+				pub fn remove<S: #scrate::GenericStorage>(
+					self,
+					storage: &S,
+					key_for: impl Fn(&#kty) -> Vec<u8>,
+				) {
+					let next_key = self.next.as_ref().map(|x| key_for(x));
+					let prev_key = self.previous.as_ref().map(|x| key_for(x));
+
+					if let Some(prev_key) = prev_key {
+						// Retrieve previous element and update `next`
+						let mut res = Self::read(storage, &*prev_key)
+							.expect("Linkage is updated in case entry is removed; it always points to existing keys; qed");
+						res.1.next = self.next;
+						storage.put(&*prev_key, &res);
+					} else {
+						// we were first so let's update the head
+						Self::write_head(storage, self.next.as_ref());
+					}
+
+					if let Some(next_key) = next_key {
+						// Update previous of next element
+						let mut res = Self::read(storage, &*next_key)
+							.expect("Linkage is updated in case entry is removed; it always points to existing keys; qed");
+						res.1.previous = self.previous;
+						storage.put(&*next_key, &res);
+					}
+				}
+
+				/// Read the contained data and it's linkage.
+				pub fn read<S: #scrate::GenericStorage>(storage: &S, key: &[u8]) -> Option<(#value_type, #linkage)> {
+					storage.get(key)
+				}
+
+				/// Generate linkage for newly inserted element.
+				///
+				/// Takes care of updating head and previous head's pointer.
+				pub fn insert_new_head<S: #scrate::GenericStorage>(
```

Comment:
Does this function need to be public? We don't write the newly created instance into the database, but already update the head + the previous head. Someone who uses this and does not know this, could break his storage?

Reply:
I've moved both `Linkage` and `Enumerator` to an internal module, so that it's clear that those structs are just implementation detail.